### PR TITLE
[minor][Feature]: Add a util and request class for fetching device token

### PR DESCRIPTION
**Pull request by ameyapat**
## PR Checklist (must be completed before review)

- [ ] All tests pass locally
- [ ] PR size is <= 500 LOC per PR Size Check policy
- [ ] PR is independently mergeable (no hidden dependencies)
- [ ] Appropriate reviewers are assigned
- [ ] PR reviewed by code owner (required if Copilot-generated)
- [ ] SME or Senior IC assigned where required

## PR Title Format

**Required Format:** `[Keyword1] [Keyword2]: Description`

- **Keyword1:** `major`, `minor`, or `patch` (case-insensitive)
- **Keyword2:** `feature`, `bugfix`, `engg`, or `tests` (case-insensitive)

**Examples:**
- `[MAJOR] [Feature]: new API`
- `[minor] [bugfix]: fix crash`
- `[PATCH][tests]:add coverage`

## Proposed changes

Add a parameter for caller to pass nonce. This will allow getting nonce as a network request separately and pass it in here from its callback. This will also prevent deallocation of initial request object during async call.

## Type of change

- [ ] Feature work
- [x] Bug fix
- [ ] Documentation
- [ ] Engineering change
- [ ] Test
- [ ] Logging/Telemetry

## Risk

- [ ] High – Errors could cause MAJOR regression of many scenarios. (Example: new large features or high level infrastructure changes)
- [ ] Medium – Errors could cause regression of 1 or more scenarios. (Example: somewhat complex bug fixes, small new features)
- [x] Small – No issues are expected. (Example: Very small bug fixes, string changes, or configuration settings changes)

### CODE REVIEW

File: IdentityCore/src/requests/MSIDDeviceTokenGrantRequest.h [hunk 1]
```diff
@@ -35,6 +35,8 @@ NS_ASSUME_NONNULL_BEGIN
```
Comment (ameyapat) [inline IdentityCore/src/requests/MSIDDeviceTokenGrantRequest.h:35]: @copilot add entry to changelog.txt to top most line below 'TBD' if top line is TBD, else add TBD as top most line. The entry should be single line that follows pattern of existing entries. The number at the end within [] is the current Pull request number: 1886

File: IdentityCore/src/requests/MSIDDeviceTokenGrantRequest.m [hunk 2]
```diff
@@ -146,31 +145,14 @@ - (instancetype _Nullable)initWithEndpoint:(nonnull NSURL *)endpoint
...
 {
-    MSIDRequestParameters *nonceReqParams = [MSIDRequestParameters new];
-    nonceReqParams.correlationId = self.context.correlationId;
-    nonceReqParams.authority = [[MSIDAADAuthority alloc] initWithURL:self.urlRequest.URL rawTenant:MSIDAADTenantTypeCommonRawValue context:self.context error:nil];
-    // Passing blank accountId details as device token is not associated with a specific account. This is required to bypass cache look up in nonce request and directly request new nonce from server.
-    nonceReqParams.accountIdentifier = [[MSIDAccountIdentifier alloc] initWithDisplayableId:@"" homeAccountId:@""];
-    MSIDNonceTokenRequest *nonceRequest = [[MSIDNonceTokenRequest alloc] initWithRequestParameters:nonceReqParams];
-    __weak typeof(self) weakSelf = self;
-    [nonceRequest executeRequestWithCompletion:^(NSString * _Nullable resultNonce, NSError * _Nullable error)
+    if ([NSString msidIsStringNilOrBlank:self.nonce])
```
Comment (kaisong1990) [inline IdentityCore/src/requests/MSIDDeviceTokenGrantRequest.m:187]: This drops the internal nonce fetch and now just errors if `.nonce` isn't already set. Looked at `MSALDeviceInfoProvider.deviceTokenWithRequestParameters` — it never sets `.nonce` before calling this, so that path will start failing with `MSIDErrorInvalidInternalParameter` as soon as this merges. Are we landing the MSAL-side change to populate the nonce in the same rollout, or does this need to go out together?

File: IdentityCore/tests/MSIDDeviceTokenGrantRequestTests.m [hunk 3]
```diff
@@ -342,6 +344,82 @@ - (void)testInit_whenSingleScope_shouldReturnNonNilRequest
     XCTAssertNotNil(request);
 }
 
+#pragma mark - executeRequestWithCompletion: nonce validation
```
Comment (kaisong1990) [inline IdentityCore/tests/MSIDDeviceTokenGrantRequestTests.m:356]: Both new tests only hit the nil/blank nonce error branch. Can we also add one that sets a real nonce and confirms it makes it into the JWT as `request_nonce`? Right now nothing actually exercises the happy path through tokenRequestWithCompletionBlock.

File: IdentityCore/src/requests/MSIDDeviceTokenGrantRequest.m (hunk 2 above)
Comment (ameyapat) [reply IdentityCore/src/requests/MSIDDeviceTokenGrantRequest.m:187]: Yes, those changes are going to be made here: https://github.com/AzureAD/microsoft-authentication-library-for-objc/pull/3032

Once this merges, I'll update common core submodule in that PR

File: IdentityCore/tests/MSIDDeviceTokenGrantRequestTests.m (hunk 3 above)
Comment (ameyapat) [reply IdentityCore/tests/MSIDDeviceTokenGrantRequestTests.m:356]: @copilot add a test where the nonce for MSIDDeviceTokenGrantRequest request is set and error is nil in the result

Comment (ameyapat) [reply IdentityCore/tests/MSIDDeviceTokenGrantRequestTests.m:356]: @copilot The test fails with error: MSIDDeviceTokenGrantRequestTests.m:473:62: fatal error: unused parameter 'result'. Please add __unused qualifier to result

**Review by antrix1989 — APPROVED** [review]

**Review by kaisong1990 — APPROVED** [review]

**Review by kaisong1990 — APPROVED** [review]

**Review by kaisong1990 — APPROVED** [review]
